### PR TITLE
Fix CalendarList rendering on react-native-web

### LINKED ISSUE #924: calendar not rendering properly with expo web
## 🐛 Bug Report

react-native-web is not working properly with react-native-calendars.

A similar issue was opened with react-native-calendars and they suggested that an issue was opened with the expo team

[Referenced comment](https://github.com/wix/react-native-calendars/issues/839#issuecomment-499484419)

Details are given about what could be the problem

> This is most probably not an issue with react-native-calendars, but with expo / webpack config. I'm using calendars fine with my web webpack setup. react-native-calendars needs transpilation (just like most react-native libraries), so I guess your webpack babel rule needs to be told to include node_modules/react-native-calendars for transpilation. If I was you, I'd close this issue and open a new one in the expo community.

### Environment

```bash
  Expo CLI 3.0.10 environment info:
    System:
      OS: macOS High Sierra 10.13.6
      Shell: 3.2.57 - /bin/bash
    Binaries:
      Node: 10.15.1 - /usr/local/bin/node
      Yarn: 1.15.2 - /usr/local/bin/yarn
      npm: 6.4.1 - /usr/local/bin/npm
      Watchman: 4.9.0 - /usr/local/bin/watchman
    IDEs:
      Xcode: 10.1/10B61 - /usr/bin/xcodebuild
    npmPackages:
      @types/react: ^16.8.23 => 16.9.2 
      @types/react-native: ^0.57.65 => 0.57.65 
      expo: ^34.0.1 => 34.0.4 
      react: 16.8.3 => 16.8.3 
      react-native: https://github.com/expo/react-native/archive/sdk-34.0.0.tar.gz => 0.59.8 
      react-navigation: ^3.11.0 => 3.11.1 
    npmGlobalPackages:
      expo-cli: 3.0.10
```

The target with the problem is `web`. `ios` and `android` are both working fine

### Steps to Reproduce

use `react-native-calendars` with `react-native-web` on `expo`

### Expected Behavior

The calendar should look like this (ios pictured)

error

### Actual Behavior

This is what it looks like on web

Screen Shot 2019-09-11 at 11 48 52 PM

The dates section is not working and the expand bar does not do anything

## Reproducible Demo

[expo snack](https://snack.expo.io/@mintgroomingapp/react-native-calendars-expo-web)

### PR DESCRIPTION
**Pull request by tedsta**
This fixes CalendarList / Agenda rendering when running on web via Expo.

Thanks to @bkniffler with his fix for touch support on web here: https://github.com/bkniffler/react-native-calendars/commit/7a0f986bbef1f84b0fd6901af6974c8bcd7436c2

The main issue was that `FlatList` scrolling only works on web if the parent component has a `flex` style. Source: https://github.com/necolas/react-native-web/issues/1436#issuecomment-612845122

Also incorporated fixes for some `PropTypes` errors on web, thanks @nandorojo: https://github.com/wix/react-native-calendars/issues/1033#issuecomment-712977254

Fixes #924

### CODE REVIEW

File: src/agenda/index.js [hunk 1]
```diff
@@ -35,7 +31,7 @@ export default class AgendaView extends Component {
     /** Specify theme properties to override specific styles for calendar parts. Default = {} */
     theme: PropTypes.object,
     /** agenda container style */
-    style: viewPropTypes.style,
+    style: PropTypes.object,
```
Comment (Inbal-Tish) [inline src/agenda/index.js:34]: 'style' prop type should be PropTypes.oneOfType([PropTypes.object, PropTypes.number, PropTypes.array]) as you can pass style inline as an object or array and you can pass it from a StyleSheet where it's just a number.

Comment (tedsta) [reply src/agenda/index.js]: Fixed

File: src/agenda/index.js [hunk 2]
```diff
@@ -15,11 +15,7 @@ import {AGENDA_CALENDAR_KNOB} from '../testIDs';
 const HEADER_HEIGHT = 104;
 const KNOB_HEIGHT = 24;
 //Fallback for react-native-web or when RN version is < 0.44
-const {Text, View, Dimensions, Animated, ViewPropTypes} = ReactNative;
-const viewPropTypes =
-  typeof document !== 'undefined'
-    ? PropTypes.shape({style: PropTypes.object})
-    : ViewPropTypes || View.propTypes;
+const {Text, View, Dimensions, Animated, Platform} = ReactNative;
```
Comment (Inbal-Tish) [inline src/agenda/index.js:18]: Why is 'import * as ReactNative from 'react-native';' still necessary? Can't we extract the components from 'react-native' without importing it all?

File: src/calendar/index.js [hunk 3]
```diff
@@ -17,11 +17,7 @@ import GestureRecognizer, {swipeDirections} from 'react-native-swipe-gestures';
 import {SELECT_DATE_SLOT} from '../testIDs';
 
 //Fallback for react-native-web or when RN version is < 0.44
-const {View, ViewPropTypes} = ReactNative;
-const viewPropTypes =
-  typeof document !== 'undefined'
-    ? PropTypes.shape({style: PropTypes.object})
-    : ViewPropTypes || View.propTypes;
+const {View} = ReactNative;
```
Comment (Inbal-Tish) [inline src/calendar/index.js:20]: Also here. Why keep importing * from 'react-native'?

File: src/agenda/index.js [hunk 4]
```diff
@@ -119,6 +115,8 @@ export default class AgendaView extends Component {
     };
 
     this.currentMonth = this.state.selectedDay.clone();
+    this.onHover = this.onHover.bind(this);
```
Comment (Inbal-Tish) [inline src/agenda/index.js:118]: What these changes has to do with fixing the rendering? I think you're mixing two tasks here...

File: src/calendar-list/index.js [hunk 5]
```diff
@@ -294,7 +291,7 @@ class CalendarList extends Component {
 
   render() {
     return (
-      <View>
+      <View style={{flex: 1}}>
```
Comment (Inbal-Tish) [inline src/calendar-list/index.js:294]: What is this for?

File: src/calendar/index.js (hunk 3 above)
Comment (tedsta) [reply src/calendar/index.js:20]: Commented above

File: src/calendar-list/index.js (hunk 5 above)
Comment (tedsta) [reply src/calendar-list/index.js:294]: This is what actually fixes the rendering - `FlatList` scrolling only works on web if the parent component has a flex style. Source: https://github.com/necolas/react-native-web/issues/1436#issuecomment-612845122

File: src/agenda/index.js (hunk 4 above)
Comment (tedsta) [reply src/agenda/index.js:118]: This adds mouse support so that the interactions can work on web. I stole this from [here](https://github.com/bkniffler/react-native-calendars/commit/7a0f986bbef1f84b0fd6901af6974c8bcd7436c2) and credited the author in the main PR comment.

File: src/agenda/index.js (hunk 2 above)
Comment (tedsta) [reply src/agenda/index.js:18]: I was trying to be as non-invasive as possible - I'm new to JS / React. Would you like me to remove the 'import * as ReactNative' and fix anything that depends on it?

Comment (Inbal-Tish) [reply src/agenda/index.js:18]: I do want to remove the * import but I don't want to break anyone. Can you tell me how do you test this in web on old RN versions?

File: src/calendar-list/index.js (hunk 5 above)
Comment (Inbal-Tish) [reply src/calendar-list/index.js:294]: If that's a style to fix a web rendering you should add it to a StyleSheet with Platform division. You can't change for all Platforms to fix one...

File: src/agenda/index.js [hunk 6]
```diff
@@ -139,6 +146,16 @@ export default class AgendaView extends Component {
     this.state.scrollY.removeAllListeners();
   }
 
+  onHover() {
```
Comment (Inbal-Tish) [inline src/agenda/index.js:149]: These are very specific implementations. If you're adding a feature, first, you should make a separate PR for that and not add it to another issue. Second, you should write an API that lets the user adjust these implementations if required.

File: src/calendar-list/index.js [hunk 7]
```diff
@@ -316,7 +313,9 @@ class CalendarList extends Component {
           showsHorizontalScrollIndicator={this.props.showScrollIndicator}
           scrollEnabled={this.props.scrollEnabled}
           keyExtractor={this.props.keyExtractor}
-          initialScrollIndex={this.state.openDate ? this.getMonthIndex(this.state.openDate) : false}
+          initialScrollIndex={this.state.openDate ? this.getMonthIndex(this.state.openDate) : 0}
+          //initialScrollIndex={1}
+          onScrollToIndexFailed={() => {}}
```
Comment (emilisb) [inline src/calendar-list/index.js:318]: Are these two lines necessary?

**Comment by FrancisLeigh** [comment]
Keen to install this after PR is merged :-)

File: src/agenda/index.js (hunk 2 above)
Comment (tedsta) [reply src/agenda/index.js:18]: Err.. I haven't tested on older RN versions. Good idea. I'll make sure that works then post back here.

File: src/agenda/index.js (hunk 6 above)
Comment (tedsta) [reply src/agenda/index.js:149]: Sure I'll split the mouse interactions into a second PR. Unless @bkniffler wants it since it is their code :) LMK

For API/adjustable-ness - none of the other interactions have this AFAICT? Am I missing something?

**Comment by nandorojo** [comment]
Thank you guys for your hard work on this, looking forward to installing it when it's merged!

File: src/calendar-list/index.js (hunk 5 above)
Comment (nandorojo) [reply src/calendar-list/index.js:294]: @tedsta Per @Inbal-Tish's comment, this should look something like:

```jsx
import { StyleSheet, Platform } from 'react-native'

...
render() {
  return <View style={Platform.select({ web: styles.webContainer, default: undefined })}>
}

const styles = StyleSheet.create({
  webContainer: {
    flex: 1
  }
})
```

File: src/agenda/index.js (hunk 6 above)
Comment (nandorojo) [reply src/agenda/index.js:149]: I don't think these hover interactions are crucial for web support. In my opinion, they should be moved to a separate PR to help get this one merged sooner.

The right way to do this would probably be to transfer over to React Native's new `Pressable`, and style it like so:

```js
<Pressable
  // react-native-web 0.14 added support for hovered here 🔥
  style={({ hovered, pressed }) => {
    opacity: pressed || hovered ? 0.7 : 1
  }}
/>
```

I'm not sure how focused on backwards-compatibility this library is, though.

**Comment by yoangau** [comment]
Hi guys, I was wondering if there was some way to patch this from our side while this issue is being resolved? Is there a way to access the necessary view to add `flex: 1` using the `theme` of `CalendarList`?

**Comment by tedsta** [comment]
Sorry guys, I'm in a new job now and don't work with React anymore 😧 Hopefully one of you guys can fork this and push it over the finish line. I'll leave this open but feel free to close.

**Comment by yoangau** [comment]
Following the last comment I opened this PR: https://github.com/wix/react-native-calendars/pull/1405

**Comment by Inbal-Tish** [comment]
@tedsta Please see merged PRs #1424  and #1405 if that solves your issue.
